src/Microsoft.Health.Fhir.SqlServer/Features/Storage/TvpRowGeneration/Merge/CompartmentAssignmentListRowGenerator.cs

```diff
@@ -0,0 +1,137 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using EnsureThat;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Models;
+using Microsoft.Health.Fhir.SqlServer.Features.Schema.Model;
+using Microsoft.Health.SqlServer.Features.Schema.Model;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Storage.TvpRowGeneration
+{
+    internal class CompartmentAssignmentListRowGenerator : ITableValuedParameterRowGenerator<IReadOnlyList<MergeResourceWrapper>, CompartmentAssignmentListRow>
+    {
+        private readonly ISqlServerFhirModel _model;
+        private readonly SearchParameterToSearchValueTypeMap _searchParameterTypeMap;
+        private bool _initialized;
+        private byte _patientCompartmentId;
+        private byte _encounterCompartmentId;
+        private byte _relatedPersonCompartmentId;
+        private byte _practitionerCompartmentId;
+        private byte _deviceCompartmentId;
+
+        public CompartmentAssignmentListRowGenerator(ISqlServerFhirModel model, SearchParameterToSearchValueTypeMap searchParameterTypeMap)
+        {
+            EnsureArg.IsNotNull(model, nameof(model));
+            EnsureArg.IsNotNull(searchParameterTypeMap, nameof(searchParameterTypeMap));
+
+            _model = model;
+            _searchParameterTypeMap = searchParameterTypeMap;
+        }
+
+        public IEnumerable<CompartmentAssignmentListRow> GenerateRows(IReadOnlyList<MergeResourceWrapper> resources)
+        {
+            EnsureInitialized();
+
+            foreach (var merge in resources.Where(_ => !_.ResourceWrapper.IsHistory)) // only current
+            {
+                var resource = merge.ResourceWrapper;
+                var typeId = _model.GetResourceTypeId(resource.ResourceTypeName);
+
+                var resourceMetadata = new ResourceMetadata(
+                    resource.CompartmentIndices,
+                    resource.SearchIndices?.ToLookup(e => _searchParameterTypeMap.GetSearchValueType(e)),
+                    resource.LastModifiedClaims);
+
+                CompartmentIndices compartments = resourceMetadata.Compartments;
```

Comment:
If "Compartments" for a single item is null, we should break the looping or move to the next item in the list?
The current code breaks, which causes other items in the list to be ignored. Is that intentional?

Reply:
This is the behavior on the previous code.

Reply:
@brendankowitz we've noticed few cases of loops breaking in the middle if a null condition is true. 
Should we instead of breaking the loop in the middle continue analyzing the remaining items in the collection?

Reply:
I replaced yield break by continue.